### PR TITLE
[Memory Improvement] delete logits before bwd - DPO style

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

https://github.com/pytorch/torchtune/pull/1235 
https://github.com/pytorch/torchtune/pull/1046

I uphold the findings from @felipemello1 in #1235; deleting logits in the LoRA DPO recipes before the backward pass, also reduces memory consumption. Isn't it just wonderful?
image

Credit to the authors of the pull requests above for the excellent suggestions. Thanks @felipemello1 for pointing this out for me to try.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

**Comment by felipemello1** [comment]
neat! Do you mind adding a comment to the code, something like "free logits otherwise it peaks backward memory". Thanks for doing it! Thats a nice memory reduction too

**Review by joecummings — APPROVED** [review]
